sql/core/src/test/scala/org/apache/spark/sql/streaming/sources/StreamingDataSourceV2Suite.scala

```diff
@@ -0,0 +1,199 @@
+  }
+
+  override def createContinuousReader(
+      schema: Optional[StructType],
+      checkpointLocation: String,
+      options: DataSourceV2Options): ContinuousReader = FakeReader
+
+  def createContinuousWriter(
+      queryId: String,
+      schema: StructType,
+      mode: OutputMode,
+      options: DataSourceV2Options): Optional[ContinuousWriter] = {
+    throw new IllegalStateException("fake sink - cannot actually write")
+  }
+
+  override def shortName(): String = "fake-both-modes"
+}
+
+class FakeStreamingNeitherMode extends DataSourceRegister with DataSourceV2 {
+  override def shortName(): String = "fake-neither-mode"
+}
+
+class StreamingDataSourceV2Suite extends StreamTest {
+
+  private def df = spark.readStream.format("rate").load()
+
+  override def beforeAll(): Unit = {
+    super.beforeAll()
+    val fakeCheckpoint = Utils.createTempDir()
+    spark.conf.set("spark.sql.streaming.checkpointLocation", fakeCheckpoint.getCanonicalPath)
+  }
+
+  testQuietly("create microbatch with only microbatch support") {
+    val query = df.writeStream.format("fake-microbatch-only").start()
+    query.stop()
+  }
+
+  testQuietly("create microbatch with both support") {
+    val query = df.writeStream.format("fake-both-modes").start()
+    query.stop()
+  }
+
+  testQuietly("create continuous with only continuous support") {
+    val query = df.writeStream
+      .format("fake-continuous-only")
+      .trigger(Trigger.Continuous(100))
+      .start()
+    query.stop()
+  }
+
+  testQuietly("create continuous with both support") {
+    val query = df.writeStream
+      .format("fake-both-modes")
+      .trigger(Trigger.Continuous(100))
+      .start()
+    query.stop()
+  }
+
+  test("microbatch with only continuous support") {
+    val ex = intercept[UnsupportedOperationException] {
+      df.writeStream.format("fake-continuous-only").start()
+    }
+
+    assert(ex.getMessage.contains(
+      "Data source fake-continuous-only does not support streamed writing"))
+  }
+
+  test("microbatch with no support") {
+    val ex = intercept[UnsupportedOperationException] {
+      df.writeStream.format("fake-neither-mode").start()
+    }
+
+    assert(ex.getMessage.contains(
+      "Data source fake-neither-mode does not support streamed writing"))
+  }
+
+  test("continuous with only microbatch support") {
+    val ex = intercept[AnalysisException] {
+      df.writeStream
+        .format("fake-microbatch-only")
+        .trigger(Trigger.Continuous(100))
+        .start()
+    }
+
+    assert(ex.getMessage.contains(
+      "Data source fake-microbatch-only does not support continuous writing"))
+  }
+
+  test("continuous with no support") {
+    val ex = intercept[AnalysisException] {
+      df.writeStream
+        .format("fake-neither-mode")
+        .trigger(Trigger.Continuous(100))
+        .start()
+    }
+
+    assert(ex.getMessage.contains(
+      "Data source fake-neither-mode does not support continuous writing"))
+  }
```

Comment:
you are testing only different types of sinks, not the different types of sources.

Reply:
Added tests for all 4*4*3 combinations of source/sink/trigger. Note that:

* I had to revert the earlier change to initialize ContinuousExecution.sources to null, because it turns out this interferes with error generation on newly constructed executions.
* Two of the cases don't throw the error until after start(). This will take a decent amount of disruptive changes to fix; the problem is that DataStreamWriter doesn't have direct visibility to what sources were used to generate it. We'd need to crawl the tree similarly to how we do it in the execution.